src/ert/_c_wrappers/util/substitution_list.py

```diff
@@ -71,6 +71,18 @@ def indexForKey(self, key):
     def substitute(self, to_substitute: str) -> str:
         return self._alloc_filtered_string(to_substitute)
 
+    def substitute_real_iter(
+        self, to_substitute: str, realization: int, iteration: int
+    ) -> str:
+        if f"<GEO_ID_{realization}_{iteration}>" in self:
+            to_substitute = to_substitute.replace(
+                "<GEO_ID>", f"<GEO_ID_{realization}_{iteration}>"
+            )
+        to_substitute = self._alloc_filtered_string(to_substitute)
+        to_substitute = to_substitute.replace("<IENS>", str(realization))
```

Comment:
After our talk I had another look at the substlist and I realized that it actually operates differently than I thought. It does substitution until fixed-point when it gets the string. This behavior was in the previous release probably broken by the introduction of substituter, but only for substitutions in `runpaths`. Basically a substitution inside a runpath is not done until fixed point. Luckily we did not introduce substituter more places, but since then it has also been used for printing `jobs.json`.

To remedy this we have to either do the `replace` until fix-point or copy the substlist, insert what we want to substitute, and let the copy do it:

```
substlist = copy.deepcopy(self)
substlist.add_item("<GEO_ID>", f"<GEO_ID_{realization}_{iteration}>")
substlist.add_item("<IENS>", str(realization))
substlist.add_item("<ITER>", str(iteration))
return substlist._alloc_filtered_string(to_substitute)
```

I am exposing the copy action to python in #4386

Reply:
I don't understand what you mean by "It does substitution until fixed-point when it gets the string"

Is there any example where your suggested code:
```python 
substlist = copy.deepcopy(self)
substlist.add_item("<GEO_ID>", f"<GEO_ID_{realization}_{iteration}>")
substlist.add_item("<IENS>", str(realization))
substlist.add_item("<ITER>", str(iteration))
return substlist._alloc_filtered_string(to_substitute)
```
does not do the same thing as the code currently in the substitution_list.py ln74-84?

Reply:
No, this substitutes `<IENS>` only at the end. I can't really come up with anything that breaks this other than contrived things, but I think that is just a lack of imagination on my part:

```
DEFINE <IT1ER> <ITER>
RUNPATH realization-<REAL>/iter-<IT<ITER>ER>/
```
For the existing code, for realization 0 iteration 1 you get runpath: `"realization-0/iter-<IT1ER>" while if you use the `add_item` code you get `realization-0/iter-1`

Reply:
Wait, looking at the subst_list again, I am not sure what it does. It seems to go through the keys only once.

Reply:
If you are a user and you have defines like:
```
DEFINE <IT1ER> <ITER>
RUNPATH realization-<REAL>/iter-<IT<ITER>ER>
```
and expect 

`realization-0/iter-1` as a result you should go to user jail :smile: 

but this will be solved by doing:

``` python 
to_substitute = to_substitute.replace("<IENS>", str(realization))
to_substitute = to_substitute.replace("<IENS>", str(realization))
to_substitute = self._alloc_filtered_string(to_substitute)
```

Reply:
Did some further testing on how the `SubstituteList` works and

```python
    subst_list = SubstitutionList()
    subst_list.addItem("<foo>", "r1")
    subst_list.addItem("<bar_r1>", "r2")
    print(subst_list.substitute("replace this <bar_<foo>>"))
```
this returns 
```python 
replace this r2
```
and this 
```python
    subst_list = SubstitutionList()
    subst_list.addItem("<bar_r1>", "r2")
    subst_list.addItem("<foo>", "r1")
    print(subst_list.substitute("replace this <bar_<foo>>"))
```
returns 
```python 
replace this <bar_r1>
```
so the order in which things are added to the substitution list matters

Reply:
Unfortunately we have seen stuff like this in the wild.